pkg/odo/cli/logs/logs.go

```diff
@@ -0,0 +1,156 @@
+package logs
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/fatih/color"
+
+	"github.com/redhat-developer/odo/pkg/log"
+
+	"github.com/redhat-developer/odo/pkg/devfile/location"
+	odoutil "github.com/redhat-developer/odo/pkg/odo/util"
+
+	"github.com/redhat-developer/odo/pkg/odo/cmdline"
+	"github.com/redhat-developer/odo/pkg/odo/genericclioptions"
+	"github.com/redhat-developer/odo/pkg/odo/genericclioptions/clientset"
+	"github.com/spf13/cobra"
+	ktemplates "k8s.io/kubectl/pkg/util/templates"
+)
+
+const RecommendedCommandName = "logs"
+
+type LogsOptions struct {
+	// context
+	Context *genericclioptions.Context
+	// clients
+	clientset *clientset.Clientset
+
+	// variables
+	componentName string
+	contextDir    string
+	out           io.Writer
+}
+
+func NewLogsOptions() *LogsOptions {
+	return &LogsOptions{
+		out: log.GetStdout(),
+	}
+}
+
+var logsExample = ktemplates.Examples(`
+	# Show logs of all containers
+	%[1]s
+`)
+
+func (o *LogsOptions) SetClientset(clientset *clientset.Clientset) {
+	o.clientset = clientset
+}
+
+func (o *LogsOptions) Complete(cmdline cmdline.Cmdline, args []string) error {
+	var err error
+	o.contextDir, err = os.Getwd()
+	if err != nil {
+		return err
+	}
+	isEmptyDir, err := location.DirIsEmpty(o.clientset.FS, o.contextDir)
+	if err != nil {
+		return err
+	}
+	if isEmptyDir {
+		return errors.New("this command cannot run in an empty directory, run the command in a directory containing source code or initialize using 'odo init'")
+	}
+
+	o.Context, err = genericclioptions.New(genericclioptions.NewCreateParameters(cmdline).NeedDevfile(""))
+	if err != nil {
+		return fmt.Errorf("unable to create context: %v", err)
+	}
+
+	o.componentName = o.Context.EnvSpecificInfo.GetDevfileObj().GetMetadataName()
+
+	o.clientset.KubernetesClient.SetNamespace(o.Context.GetProject())
+
+	return nil
+}
+
+func (o *LogsOptions) Validate() error {
+	return nil
+}
+
+func (o *LogsOptions) Run(ctx context.Context) error {
+	containersLogs, err := o.clientset.LogsClient.DevModeLogs(o.componentName, o.Context.GetProject())
+	if err != nil {
+		return err
+	}
+
+	for container, logs := range containersLogs {
+		err = printLogs(container, logs, o.out)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// printLogs prints the logs of the containers with container name prefixed to the log message
+func printLogs(containerName string, rd io.ReadCloser, out io.Writer) error {
+	reader := bufio.NewReader(rd)
+	var lines []string
+	for {
```

Comment:
Could we use a scanner instead?

https://pkg.go.dev/go/scanner

Reply:
What are the advantages of using `scan.Scanner` over `bufio.Reader`?

Reply:
Scanner handles different end of lines for you

Reply:
And as far as logs are concerned, what could those different end of lines be? Even in general, what other end of lines does one need to consider? Genuinely curious.

Reply:
Unix systems use `\n` when Windows ones use `\r\n`

I'm not sure what could be returned as part of the logs, probably `\n` only, but not sure. `Scanner` helps not bother with this.